cmd/xl-v1-healing-common.go

```diff
@@ -261,38 +248,39 @@ func xlHealStat(xl xlObjects, partsMetadata []xlMetaV1, errs []error) HealObject
 // - errs updated to have errFileNotFound in place of disks that had
 // missing parts.
 // - non-nil error if any of the online disks failed during
-// calculating blake2b checksum.
-func disksWithAllParts(onlineDisks []StorageAPI, partsMetadata []xlMetaV1, errs []error, bucket, object string) ([]StorageAPI, []error, error) {
+// calculating bitrot checksum.
+func disksWithAllParts(onlineDisks []StorageAPI, partsMetadata []xlMetaV1, errs []error, bucket,
+	object string) ([]StorageAPI, []error, error) {
+
 	availableDisks := make([]StorageAPI, len(onlineDisks))
-	buffer := healBufferPool.Get().(*[]byte)
-	defer healBufferPool.Put(buffer)
+	buffer := []byte{}
 
-	for diskIndex, onlineDisk := range onlineDisks {
+	for i, onlineDisk := range onlineDisks {
 		if onlineDisk == OfflineDisk {
 			continue
 		}
 		// disk has a valid xl.json but may not have all the
 		// parts. This is considered an outdated disk, since
 		// it needs healing too.
-		for _, part := range partsMetadata[diskIndex].Parts {
+		for _, part := range partsMetadata[i].Parts {
 			partPath := filepath.Join(object, part.Name)
-			checkSumInfo := partsMetadata[diskIndex].Erasure.GetChecksumInfo(part.Name)
-			hash := checkSumInfo.Algorithm.New()
-			_, hErr := io.CopyBuffer(hash, StorageReader(onlineDisk, bucket, partPath, 0), *buffer)
-			if hErr == errFileNotFound {
-				errs[diskIndex] = errFileNotFound
-				availableDisks[diskIndex] = OfflineDisk
-				break
-			}
-			if hErr != nil && hErr != errFileNotFound {
+			hashInfo := partsMetadata[i].Erasure.GetChecksumInfo(part.Name)
+			verifier := NewBitrotVerifier(hashInfo.Algorithm, hashInfo.Hash)
+
+			// verification happens even if a 0-length
+			// buffer is passed
+			_, hErr := onlineDisk.ReadFile(bucket, partPath, 0, buffer, verifier)
+			if hErr != nil {
+				_, isCorrupted := hErr.(hashMismatchError)
+				if isCorrupted || hErr == errFileNotFound {
```

Comment:
Comment for this function needs to be updated, currently it says:
```
Returns,                                                                                                                           
// - disks which have all parts specified in the latest xl.json.                                                                                                                
// - errs updated to have errFileNotFound in place of disks that had                                                                                                            
// missing parts.                                                                                                                                                               
// - non-nil error if any of the online disks failed during                                                                                                                     
// calculating bitrot checksum.
```

Reply:
Fixed the last point and re-formatted the comments - we only error if the disks had unexpected errors.